### PR TITLE
Remove OldAlpha- Bench: 5967663

### PR DESCRIPTION
**Pull request by VoyagerOne**
Simplification:  Remove Old Alpha.
It may not be wise to save leaf positions in TT without trying a move.

Since the position may be drawn or zugzwang. 

STC:
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 49481 W: 8995 L: 8923 D: 31563

LTC:
LLR: 2.96 (-2.94,2.94) [-3.00,1.00]
Total: 74246 W: 9703 L: 9652 D: 54891

Bench: 5967663

### CODE REVIEW

**Comment by snicolet** [comment]
I would recommend testing for non-regression with multiple threads too (something like 5+0.05, 11 threads, 64 hash, either SPRT or a fixed number of games).

The reason I suggest this is because the transposition table is the primary tool we use in LazySMP to communicate between threads, and adding or removing more positions in the tt with flag BOUND_EXACT by one thread might prove critical for the pruning of nodes by other threads (line 620 of search.cpp). One example where I was surprised by this effect is the following patch, which was yellow with one thread and disastrous with 11 threads:

about +0.7 Elo on one thread:
http://tests.stockfishchess.org/tests/view/58f68d520ebc59035df33c99

about -11.8 Elo on eleven threads:
http://tests.stockfishchess.org/tests/view/58f723fd0ebc59035df33cb9

I know that multi-threads tests are long, but better safe than sorry with the SMP implementation! :-)

**Comment by VoyagerOne** [comment]
@snicolet 
Your patch has a major flaw... which is why it failed badly for SMP.

In your patch you are saving to the TT before searching the remaining moves for that node.
So on move1 you may identify the node as BOUND_EXACT.
But on move7  a cutoff occurs so the node is not an Exact Bound but a LOWER Bound.
Meanwhile all the other threads believe the node is EXACT.

You should only update the TT for a node when all the moves are exhausted or a cutoff occurs.

**Comment by lucasart** [comment]
That's not a simplification, it's a bug.

**Comment by zamar** [comment]
@VoyagerOne: I do understand the old code, but I do not understand your change.

**Comment by VoyagerOne** [comment]
@lucasart @zamar   
Actually this patch fixes a bug...

In the current master...
You are assuming that at the node (with no "available moves") it's an Exact_Bound.
You cannot assume this without testing all the moves... in QS we only select a small subset of all available moves (such as recapture).

For all we know if we played a regular quiet at that node it may cause a cut-off.

**Comment by zamar** [comment]
@VoyagerOne: 
- At QS the side to move can choose between static eval and captures. If none of the captures work, then the best is the static eval.
- If we reach the same position again, we can read this result from the transposition table (without having to retest all the captures again which still are not going to work...).

I can't see the bug...

**Comment by VoyagerOne** [comment]
@zamar 
Right I understand that...but this not what OldAlpha is doing.

OldAlpha is used if there are no moves available

So you are at a node.
You get the staticEval.
You generate recapture moves at that node.
However generate comes back with no moves.
You then check to see if static staticEval > oldAlpha.
If it is you identify the node as Exact.

This logic is flaw...since you can't determine if the node is exact without playing any moves.

This bug will more pronounce in SMP

**Comment by zamar** [comment]
@VoyagerOne: I can't see a flaw in the logic you describe.

If staticEval > oldAlpha and staticEval < beta and none of the available moves (captures) can improve, then the result "value == staticEval" is exact.

**Comment by VoyagerOne** [comment]
@zamar 
What you said is correct...

Now let me rephrase your statement

If staticEval > oldAlpha and staticEval < beta and **!inCheck and no moves can be played (i.e. draw)**, then the result "value == staticEval" is exact.

This is where the bug comes into play...not when we run out of moves at the node...but when there are NO moves that can be played at that node.

This could well be the reason why SF scored 26.0  and dropped to 0 during that infamous TCEC game.

**Comment by zamar** [comment]
@VoyagerOne: I don't follow. If there are moves available, we do examine them in qsearch, verify that they can't improve staticEval, and then reason that the result (staticEval) is exact.

That is fully intentional and correct.

**Comment by VoyagerOne** [comment]
@zamar 
Sorry if I am not explaining this well...but it is a bit difficult to understand / explain.

Let me try an example:

Note that oldAlpha is not a stand pat value.  It gets recursively copied down for each ply in qs.

//Initialize the node
Say we are at depth = -5
oldAlpha = -30
bestValue = staticEval = 80

// move loop
   // all moves are terrible with the best value being -50
   // bestValue never gets updated and no bestMove found
// done move Loop

we now save to TT
The actual value for this node is -50, so we should update it as so.

But since 
bestValue > oldAlpha
We save it as an Exact_Bound with a score of 80.
We save bestMove as Move_None, since we know that none of the moves tried can score 80 (best is -50)

**Comment by VoyagerOne** [comment]
Also see the SMP results:
http://tests.stockfishchess.org/tests/view/590382a50ebc59035df3409f

I am aware of the large error margin...
but given enough games, it will show clear superiority.

As I mentioned above the bug will be more pronounce in SMP

**Comment by MichaelB7** [comment]
This patch appears to be sound in my testing and I believe it does address a bug.

**Comment by zamar** [comment]
- The example you posted explain how the system works. And in my understanding the behaviour is fully correct.
- SMP results are inconclusive.
- Sorry, but unless you can provide a clear example of how the system would misbehave with the current implementation, I can't commit this as a bugfix. All examples that you've posted so far describe how the system currently works, but they do not point out why that behaviour would be incorrect.

**Comment by jhellis3** [comment]
Because we are creating false lower bounds?  I don't see how a false lower
bound can be correct, and if it is not correct then.... what else could it
be?

**Comment by DU-jdto** [comment]
The value is exact because any (alpha, beta) window with alpha < value < beta will result in that value being selected. BOUND_LOWER and BOUND_UPPER are to indicate 'this is the value we found, but we have reason to believe that we might have found a different value if we'd searched with a wider window'.

**Comment by jhellis3** [comment]
No.....

**Comment by zamar** [comment]
@jhellis3: Why would the lower bound be false?

- In qsearch, staticEval always guarantees you the lower bound. You can always stop.
- We tried all the available captures and none of them worked, so the value is also the upper bound of what we can get in qsearch.

- If the value is both the upper bound and lower bound, it's the exact value.

**Comment by jhellis3** [comment]
Alpha is still the lower bound if you do not have any good moves.... I mean if you can't play a move that raises alpha how can you say you have a lower bound higher than it. It makes no sense.

I agree, one can certainly use static eval as an _upper_ bound (but we are currently using it as both). 

What we should probably be doing is just returning alpha before the tt save if bestMove == MOVE_NONE.  Or set bestValue = alpha in that case, at which point BOUND_EXACT would make sense again.

**Comment by Vizvezdenec** [comment]
For me it looks like a clear simplification, it removes some code and logic and passed [-3;1] SPRT fully, why can't this be commited as it?

**Comment by jhellis3** [comment]
The current code actually only addresses half the problem though AFAICS (the incorrect lower bound).  Even if we just use UPPER_BOUND in that case, the value we are saving to the TT is still not quite accurate IMHO.

Adding:
if (bestMove == MOVE_NONE)
        bestValue = alpha;

just before the tt save is probably most accurate... this will also solve the bounds issue since alpha will never be greater than oldAlpha unless there is an actual move which raises it.

**Comment by DU-jdto** [comment]
@jhellis3 The whole basis of standing pat is that it makes sense to treat the static evaluation as a lower bound in the qsearch, even if none of the moves searched exceed its value (the logic being that at least one of the quiet moves will likely improve the position, as zugzwang is very rare).

**Comment by jhellis3** [comment]
That is just raising alpha with no basis for doing so though.... it makes no sense.

**Comment by DU-jdto** [comment]
The basis is what I just said, that at least one of the quiet moves will likely improve the position (and actually searching those quiet moves is of course out of the question, as the qsearch mustn't slow down the search too much). It's less precise than a full search, of course, but that's an intentional tradeoff. The rare zugzwang that trips up the stand pat assumption will be located in a later search iteration when there's more depth available (and if you don't have time for another iteration, well, low depth searches sometimes miss things, that's nothing new).

**Comment by jhellis3** [comment]
Zugzwang is not the issue.... but I digress....

**Comment by DU-jdto** [comment]
It occurs to me that there _is_ a slight TT reliability issue here in that 'no checks' searches and 'recaptures' searches are saved at the same TT depth. I assume this was tested at some point and found not to lose ELO.

**Comment by snicolet** [comment]
@zamar 

I think the current code is correct, provided that we generate the same (possibly empty) set of moves when we come back to the same qsearch() call later with the same position and depth.

However, as I understand it, @VoyagerOne is trying to tell us that this assumption is not always true when the last move is at depth <= -5, because of the way our move picker changes its generation policy in line 103 of MovePicker.cpp

For example, consider the following lines from a position P with White to play :

• line A :   Nxf4, Qxb2, Rxh7
• line B :   Rxh7, Qxb2, Nxf4

After line A, it is Black to play and she generates the (possibly empty) set of recaptures on square h7 (let's call this set `movesA`)
After line B, it is also Black to play and she generates the (possibly empty) set of recaptures on square f4 (let's call this set `movesB`).

Yet the positions after lines A and B are the same (so they have the same `posKey`in line 1193 of search.cpp), and they are searched at the same depth in qsearch.

VoyagerOne's claim is that we should not store an exact bound in the transposition table after line A when `moveA` is empty, because we may probe the transposition table (and make a wrong pruning decision) when come back later after line B where `movesB`might be non-empty.

I think VoyagerOne has a point here (and that @DU-jdto has the same though, reading his last comment) about the current code being somewhat risky.

However, it is not clear to me that the 'fix' in the pull request is the proper answer to this issue... Basically, when `movesA` and `movesB`are different sets of moves, then we shouldn't use the same transposition table slot to make pruning decisions :-(

**Comment by DU-jdto** [comment]
@snicolet This pull request doesn't help at all. The problem mentioned doesn't call into question the value's status as a lower bound, only as an _upper_ bound.

**Comment by DU-jdto** [comment]
@snicolet I'm pretty sure the solution is what I posted above, to never set BOUND_UPPER at 'recaptures' depths. The trouble is, this might still lose ELO, as the extra pruning may be a net gain in ELO even if it isn't strictly speaking safe. I just wonder whether it's been tested before.

**Comment by zamar** [comment]
@snicolet: What you describe is a gap in the current TT-handling. However this patch doesn't really address the problem.

I tried to fix the problem you describe a couple of years back, but at least at the time it was an ELO loss...

**Comment by VoyagerOne** [comment]
I think what I am trying to get at.
Is that you can NOT identify a qs leaf node, without trying any sibling nodes.
The only safe way to identify a node without any siblings is if the node is a terminal node.  (Checkmate or Draw)

A qs leaf node is a node that we no longer generate moves...if you save that nodeType in the TT it will provide misinformation.

**Comment by VoyagerOne** [comment]
And as I mentioned before... this will have a bigger impact on SMP than single thread.
Very similar to @snicolet test posted above.

**Comment by zamar** [comment]
I'm rejecting the patch because in my view it's neither a bugfix or an honest simplification.

If it can be shown to be an ELO gain (even in SMP case), I'm happy to commit, but I strongly doubt that this would be a case.

**Comment by snicolet** [comment]
@DU-jdto: I have pushed a [0..4] test which tries to use different TT slots for different recaptures moves when we are at recaptures depths. This is a radical way of handling the problem, which has advantages (simplicity) and disadvantages (less sharing of information, in particular for ttMove). http://tests.stockfishchess.org/tests/view/5905bf400ebc59035df34116

If you think that your solution (not using BOUND_UPPER at recapture depths) might work and bring an Elo gain, feel free to test it, of course :-)

**Comment by AlexandreMasta** [comment]
Come on. Just increase the number of games of the test made or make a new test with predefined number of games. Period, you have your answer.

**Comment by AlexandreMasta** [comment]
@zamar This patch is an elo gain.

**Comment by zamar** [comment]
It needs to be proved.
